src/OpenTelemetry/Metrics/MetricPointOptionalComponents.cs

```diff
@@ -38,4 +40,30 @@ internal MetricPointOptionalComponents Copy()
 
         return copy;
     }
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    public void AcquireLock()
+    {
+        if (Interlocked.CompareExchange(ref this.isCriticalSectionOccupied, 1, 0) != 0)
```

Comment:
lets add some code comments to help with intuition/future readers?
something like:

isCriticalSectionOccupied = 0,1 indicates if some thread is already occupied or not.
The CAS is to do the following **atomically**:

If the original value was 0 (i.e unoccupied), set it to 1 (i.e occupied) and return (successfully acquired lock)
If the original value was 1 (i.e already occupied), don't change it, but spin and keep trying indefinitely until we succeed.